### ISSUE TITLE
[Feature Request] Revise Backend User permissions and Backend User groups, allow for hierarchical permissions and grouping.

### ISSUE THREAD

**Issue by wilirius** [octobercms/october#2367]
Right now in October the only user type which can create other users is a superuser, and a user given the permission to "Manage other administrators."

The problem is that permissions are nearly completely flat with the exception that only superusers can create superusers.

The "Manage other administrators" permission allows the user to edit ANY other user (even superusers, even their passwords), put anyone in any group, and they can create another user with permissions that it doesn't even have.

It would be a much more useful system if the groups and permissions were hierarchical. This would mirror the power structures that already exist in many CMSs and organized groups of people throughout the world.

Some ideas I am thinking of:
1. Groups needs to be split into 2 taxonomies (roles, groups), as it currently serves the purpose of 1-categorizing and 2- grouping permissions.
2. Roles will allow grouping of permissions.
3. A user has to be assigned a role (possibly only 1).
4. Make roles have rank (superuser, admin, subadmin, etc).
5. A user cannot give permissions at or above those defined by its role.
6. A user cannot add a user to its same role, only to roles below its own.
7. Groups will allow for the organizing of groups of people.
8. A user can only add users to its own group.

**Comment by LukeTowers** [comment]
I think a good start to this might be to simply check if the logged in backend user has the permission it's attempting to assign to a user before actually assigning it. I agree that more granular permissions as they relate to managing groups are definitely necessary.

**Comment by wilirius** [comment]
That's a great idea. If it would also be possible to assign ranking values to the groups of permissions, then that would solve a lot of the rest of the list. The main thing is I don't know if some people have used the current group implementation for reasons other than groups of permissions, and if so, this could mess with their expected functionality.

**Comment by leocavalcante** [comment]
That was one of my first complaints about OctoberCMS too.
https://github.com/leocavalcante/oc-backendusertree

**Comment by daftspunk** [comment]
This will need to go through a discovery process. Whatever solution that is added to the core needs to remain as simple as the current. The proposal from @wilirius does appear to make the most sense so far.

**Comment by bbrala** [comment]
We currently develop in Drupal and Laravel and was talking to one of our Drupal devs about this to see how we manage our roles and permissions there.

The current solution we use there is as follows.

1. A role is a collection of permissions
1. Users are assigned roles
1. Every role creates a permission for a role to be able to assign that a specific role to a user. (eg. `permission to assign the newseditor role to users`
1. Admins (superusers) can only define roles.

This way the admins can create role's and even define which roles can assign a specific role. Permissions are never specifically assigned to users, but it is purely role based.

So a usecase would be; the superuser creates a role 'news editor' with permissions to edit and publish news. The administrator created a role 'manager of news' with the permission to 'assign the role news editor to a user'. They then assign the role 'news editor' and 'manager of news' to the head of the news.

The reason we use this approach is that we want to give the users the ability to assign roles, but we do not want to enable most users to create roles, since that would mostly mean a mess. 

You could say that this approach is maybe overkill in some simple situations. But could also argue that if that is the case you could just have the client be superuser or a role with a most permissions.

**Comment by daftspunk** [comment]
This feature is implemented. Only super admins can define and assign roles for the moment. Users still have granular permissions because of an additional concept called "System roles", allowing permissions to be modified on a user-by-user basis.

System roles are roles defined by a special code, October will ship with 2 system roles:

- Developer
- Publisher

Publisher, for example, will not have access to the CMS area or the Builder plugin, since these are developer-centric tools. This should make it easier to quickly create client accounts.

The main difference is

- user defined roles include nothing by default
- system roles include everything by default

So if a new permission is added to the application, say by installing a new plugin. Users who have system roles will be granted the new permission by default, whereas user roles will need to be updated to explicitly grant the newly included permission.

As above, when a permission is defined, it will be included in all system roles by default. However, a permission can specify the "role API codes" it wishes to belong to.

                'cms.manage_pages' => [
                    'label' => 'cms::lang.permissions.manage_pages',
                    'tab' => 'cms::lang.permissions.name',
                    'roles' => ['developer', 'foobar'],
                    'order' => 100
                ],

In the above example, the `cms.manage_pages` permission will be enabled for roles that use the API code "developer" or "foobar" -- these are now reserved codes and will cause the role to have it's permissions locked. Since this permission defines roles, all other system roles will *not* have this permission enabled.

In other words, if a permission declares the DEVELOPER role, then the PUBLISHER role will not receive this permission.

See this commit for more details https://github.com/octobercms/october/commit/4fd1ca824f7873b540d66d8b57c7e40b2308f172
